cli/src/commands/parse/msgs.rs

```diff
@@ -0,0 +1,426 @@
+const STREAM_PATH_ATTACHMENT_STORE_PREFIX: &str = "__attach_version1.0_#";
+
+static CONTENT_TYPE_MIME_HEADER_RX: Lazy<Regex> =
+    Lazy::new(|| Regex::new(r"Content-Type:((\s)+.+\n)+").unwrap());
+static CONTENT_TRANSFER_ENCODING_MIME_HEADER_RX: Lazy<Regex> =
+    Lazy::new(|| Regex::new(r"Content-Transfer-Encoding:((\s)+.+\n)+").unwrap());
+static STREAM_PATH_MESSAGE_BODY_PLAIN: Lazy<PathBuf> =
+    Lazy::new(|| PathBuf::from("__substg1.0_1000001F"));
+static STREAM_PATH_MESSAGE_HEADER: Lazy<PathBuf> =
+    Lazy::new(|| PathBuf::from("__substg1.0_007d001F"));
+static STREAM_PATH_ATTACHMENT_FILENAME: Lazy<PathBuf> =
+    Lazy::new(|| PathBuf::from("__substg1.0_3707001F"));
+static STREAM_PATH_ATTACHMENT_EXTENSION: Lazy<PathBuf> =
+    Lazy::new(|| PathBuf::from("__substg1.0_3703001F"));
+static STREAM_PATH_ATTACHMENT_DATA: Lazy<PathBuf> =
+    Lazy::new(|| PathBuf::from("__substg1.0_37010102"));
+
+pub struct Statistics {
+    processed: AtomicUsize,
+    failed: AtomicUsize,
+    uploaded: AtomicUsize,
+}
+
+impl Statistics {
+    fn new() -> Self {
+        Self {
+            processed: AtomicUsize::new(0),
+            failed: AtomicUsize::new(0),
+            uploaded: AtomicUsize::new(0),
+        }
+    }
+
+    #[inline]
+    fn add_uploaded(&self, num_uploaded: usize) {
+        self.uploaded.fetch_add(num_uploaded, Ordering::SeqCst);
+    }
+
+    #[inline]
+    fn increment_failed(&self) {
+        self.failed.fetch_add(1, Ordering::SeqCst);
+    }
+
+    #[inline]
+    fn increment_processed(&self) {
+        self.processed.fetch_add(1, Ordering::SeqCst);
+    }
+
+    #[inline]
+    fn num_uploaded(&self) -> usize {
+        self.uploaded.load(Ordering::SeqCst)
+    }
+
+    #[inline]
+    fn num_failed(&self) -> usize {
+        self.failed.load(Ordering::SeqCst)
+    }
+
+    #[inline]
+    fn num_processed(&self) -> usize {
+        self.processed.load(Ordering::SeqCst)
+    }
+}
+
+#[derive(Debug, StructOpt)]
+pub struct ParseMsgArgs {
+    #[structopt(short = "d", long = "dir", parse(from_os_str))]
+    /// Directory containing the msgs
+    directory: PathBuf,
+
+    #[structopt(short = "s", long = "source")]
+    /// Source name or id
+    source: SourceIdentifier,
+
+    #[structopt(long = "transform-tag")]
+    /// Transform tag to use.
+    transform_tag: TransformTag,
+
+    #[structopt(short = "n", long = "no-charge")]
+    /// Whether to attempt to bypass billing (internal only)
+    no_charge: bool,
+
+    #[structopt(short = "y", long = "yes")]
+    /// Consent to ai unit charge. Suppresses confirmation prompt.
+    yes: bool,
+}
+
+fn read_stream(stream_path: PathBuf, compound_file: &mut CompoundFile<File>) -> Result<Vec<u8>> {
+    let data = {
+        let mut stream = compound_file.open_stream(stream_path)?;
+        let mut buffer = Vec::new();
+        stream.read_to_end(&mut buffer)?;
+        buffer
+    };
+
+    Ok(data)
+}
+
+fn read_unicode_stream_to_string(
+    stream_path: PathBuf,
```

Comment:
Here and in a few other places you don't actually need a `PathBuf`, just a `Path`, which allows you to remove a bunch of `clone()`s elsewhere.